### PR TITLE
WIP Feature/init

### PR DESCRIPTION
**Pull request by cescoferraro**
This aims to reduce boilerplate needed to bootstrap common projects. Will start by Golang.
The template itself needs more work. 

Right now I am getting `found unexpected end of stream` at the last line of the template.

### CODE REVIEW

**Comment by dnephin** [comment]
This is an interesting idea. It reminds me of https://github.com/audreyr/cookiecutter

We could maybe create some templates for `cookiecutter` and suggest using it from the docs?

If we add our own `init`, I think the templates should be external. Maybe it could use `gist` or `github` to fetch them? Keeping them external will make it easier for people to use their own without having to add a bunch of templates to `dobi`.

**Comment by cescoferraro** [comment]
I was thinking on a gist. Then it would be easy to assing a flag to set your custom gist url

`dobi init golang --gist http:/....`

**Comment by cescoferraro** [comment]
I push though the IDE and it somehow messed stuff on my commits...
ultimate

**Comment by dnephin** [comment]
I think this will be cool. I'm going to look into it more this weekend. I'd like to be able to do `dobi init `dnephin/go-template` or something like that, when I create a new repo.

What do you mean by your IDE messes up your commits?

File: cmd/golang.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+type TemplatingInfo struct {
+	Name string
+	Path string
+}
+
+func newGolangCommand(opts *dobiOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "golang",
+		Short: "Run the remove action for all resources",
```
Comment (dnephin) [inline cmd/golang.go]:
```go
Use: "init",
Short: "Generate a dobi.yaml from a template"
```

File: cmd/golang.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			gopath := os.Getenv("GOPATH")
```
Comment (dnephin) [inline cmd/golang.go]: I like to keep the anonymous run functions really short. It also helps with indentation and keeping shorter lines.

Could you have this call a top level function (like the other commands) ?

**Comment by dnephin** [comment]
I'm back to wondering if the best way to implement this would be just some pattern for running `cookiecutter` (or something similar) in a container.

**Comment by cescoferraro** [comment]
https://github.com/cescoferraro/go-dobi-cutter

This works
`docker run -it   -v $PWD:/src/app -w /srv/app python:2 pip install cookiecutter && cookiecutter https://github.com/cescoferraro/go-dobi-cutter.git`


This works but the project folder get owned by root
mount=source:
    bind: .
    path: /srv/app

image=python:
    image: python
    tags: ["3"]

job=golang-init:
    use: python
    mounts: [source]
    interactive: true
    working-dir: /srv/app
    command: bash -c -E "cookiecutter https://github.com/cescoferraro/go-dobi-cutter.git"

**Comment by dnephin** [comment]
Ya, that's an issue we have to deal with for a lot of tasks. We could assume user `1000`, but that won't always be correct. I don't know of any way to deal with that without an entrypoint script that creates the user with the passed in uid.

**Comment by cescoferraro** [comment]
This happens here too with things like `dobi deps`, but could be easily solved with user: `"{user.uid}"`.
But pip needs more permissions to run

**Comment by cescoferraro** [comment]
Just started the cookie cutter way

File: cmd/init.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	return nil
+}
+
+func startCookieCutterContainer(client docker.Client) (docker.Container, error) {
+	pwd, err := os.Getwd()
+	if err != nil {
+		return docker.Container{}, err
+	}
+	dockeropts := docker.CreateContainerOptions{
```
Comment (dnephin) [inline cmd/init.go]: It's kind of unfortunate that we now have this "container run" code in two places, one for the job task and one for init.

I'm wondering if we could re-use `dobi`s ability to run containers to make this work.

If we added a `--url` flag (similar to `-f`) to point `dobi` at a remote `dobi.yaml`, we could do something like this:

```
dobi --url https://raw.githubusercontent.com/cescoferraro/go-dobi-cutter/master/dobi.yaml
```

This solution works really well with [the idea](https://github.com/dnephin/dobi/issues/22#issuecomment-257202850) for allowing`include` to also be remote urls. The code should be about the same, so can be mostly re-used.

It makes it easy for anyone to create their own init, or other bootstrapping process, it doesn't even have to use cookiecutter.

The only issue I see is that the github urls are a bit long, but that can be solved pretty easily with either url shortners, or with a better host for the `dobi.yaml` files.

File: config/job.go [hunk 4]
```diff
@@ -73,13 +73,13 @@ type JobConfig struct {
...
-	// type: list of 'host_port:container_port'
-	Ports []string
-	// Devices Maps the host devices you want to connect to a container
+	// Devices  Maps the host devices you want to connect to a container
 	// type: list of device specs
 	// example: ``{Host: /dev/fb0, Container: /dev/fb0, Permissions: rwm}``
 	Devices []Device
+	// Ports Publish ports to the host
+	// type: list of 'host_port:container_port'
+	Ports []string
```
Comment (dnephin) [inline config/job.go]: It looks like some unrelated changes about devices got mixed in here.

**Comment by cescoferraro** [comment]
I will start off a new branch. I can implement the url flag but I dont think we need it. Instead we could just validate whenever or not the --filename flag is a URL. Then we could issue the req, get the responseś slice of bytes and toss into LoadFromBytes. I can submit a PR.

But that does not solve the init, that conceptually has to create some files on the fs. We should run init with the same approach of passing everything through the dobi's engine, You are right, we should be DRY.

But we would still need a tool to actually templating those files. cookiecutter does seem like a a nice fit.
THe #55 would help here to make sure this the dobi-cookiecutter job runs with as the right user/permissions. 

The init routine shoud create a  generic dobifile like

```
meta:
    include: ["{{ .dirname }}/*.yaml"]
```


Get the []bytes from a template ilke the below, and pass the result through LoadFromBytes
```
var initModel :=`
mount=source:
    bind: .
    path: /srv/app

image=cookkie:
    image: init-image
    steps:
      - FROM: python:3
      - RUN: pip install cookiecutter
      // set a user somehow
      - RUN: user{{.user}}:{{.uid}}

job=init:
    use: cookkie
    mounts: [source]
    interactive: true
    working-dir: /srv/app
    command:  cookiecutter {{.cookieFlag}}
`
```
and run the cookicutter task to create the {{.dirname}} folder of a repo we gould pass like

`dobi init --cookie github.com/me/cutter`

What do you think?

**Comment by dnephin** [comment]
Sorry I haven't responded here in a while. I've thought a bit more about this init problem.  Ultimately I don't think it's a `dobi` concern. `dobi` should strive for a concise config, but initialising a new repo is best left to other tools.

`dobi` is only used as a way to "package" those tools with a docker image and `run` configuration. `dobi` isn't specifically designed as a package manager. https://github.com/bfirsh/whalebrew is much better suited to that problem.

So an ideal way to provide repo initialization might be with a whalebrew package for cookiecutter, and some cookiecutter templates for different languages, which include a `dobi.yaml`.

However, with the `--url` option on `dobi` it should still be possible to do something similar, but I don't think we should add any features specifically for this use case. We could even do it with the existing `-f` flag, it just wouldn't support remote configs yet.

Once we have `--url` we can add to the docs about how to use it to initialize repos, by pointing at example configs. If you're interested in submitting a PR for this, it would be greatly appreciated.

Thanks again for bringing up this idea!
